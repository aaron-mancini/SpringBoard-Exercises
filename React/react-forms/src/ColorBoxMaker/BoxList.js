import React, { useState } from 'react';
import NewBoxForm from './NewBoxForm';
import Box from './Box';
import { v4 as uuidv4 } from "uuid";

const BoxList = () => {
    let [boxes, setBoxes] = useState([])

    const addBox = box => {
        let newBox = {...box, id: uuidv4() }
        setBoxes(boxes => [...boxes, newBox])
    }

    const removeBox = boxId => {
        let newBoxes = boxes.filter(b => b.id !== boxId)
        setBoxes(newBoxes);
    }

    return (
        <>
            <NewBoxForm addBox={addBox}/>
            {boxes.map(b => <Box key={b.id} backgroundcolor={b.color} height={b.height} width={b.width} removeBox={removeBox} id={b.id}/>)}
        </>
    )
}

export default BoxList;