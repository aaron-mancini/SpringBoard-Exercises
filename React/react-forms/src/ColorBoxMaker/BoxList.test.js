import React from "react";
import { render, fireEvent } from "@testing-library/react";
import BoxList from "./BoxList";

it("renders without crashing", function() {
    render(<BoxList />);
});

it("matches snapshot", function() {
    const { asFragment } = render(<BoxList />);
    expect(asFragment()).toMatchSnapshot();
});

it("can add a new box", function() {
    const { getByLabelText, queryByText } = render(<BoxList />)

    expect(queryByText("X")).not.toBeInTheDocument();

    const colorInput = getByLabelText("Color:");
    const submitBtn = queryByText("Add a new Box!");

    fireEvent.change(colorInput, { target: { value: "red" }})
    fireEvent.click(submitBtn);

    expect(queryByText("X")).toBeInTheDocument();
    expect(queryByText("X").previousSibling).toHaveStyle(`
        width: 100px;
        height: 100px;
        background-color: red;
    `)
})

it("can remove a box", function() {
    const { getByLabelText, queryByText } = render(<BoxList />)

    const colorInput = getByLabelText("Color:");
    const submitBtn = queryByText("Add a new Box!");

    fireEvent.change(colorInput, { target: { value: "red" }})
    fireEvent.click(submitBtn);

    const removeBtn = queryByText("X");

    fireEvent.click(removeBtn)
    
    expect(removeBtn).not.toBeInTheDocument();
})