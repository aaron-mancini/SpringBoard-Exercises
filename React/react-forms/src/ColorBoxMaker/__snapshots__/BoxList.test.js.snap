// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <form>
    <label
      for="color"
    >
      Color:
    </label>
    <input
      id="color"
      name="color"
      value=""
    />
    <label
      for="height"
    >
      Height:
    </label>
    <input
      id="height"
      name="height"
      type="number"
      value="100"
    />
    <label
      for="width"
    >
      Width:
    </label>
    <input
      id="width"
      name="width"
      type="number"
      value="100"
    />
    <button>
      Add a new Box!
    </button>
  </form>
</DocumentFragment>
`;
